src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java

```diff
@@ -142,4 +144,122 @@ public void close() throws IOException {
         ch.close();
     }
 
+    private static final int TRANSFER_SIZE = 8192;
+
+    @Override
+    public long transferTo(OutputStream out) throws IOException {
+        Objects.requireNonNull(out, "out");
+
+        if (out instanceof ChannelOutputStream cos) {
+            return transferInBlockingMode(this.ch, cos.channel());
+        }
+
+        return super.transferTo(out);
+    }
+
+    private static long transferInBlockingMode(ReadableByteChannel rbc, WritableByteChannel wbc) throws IOException {
+        return supplyUsingBlockingChannel(rbc, () -> supplyUsingBlockingChannel(wbc, () -> doTransfer(rbc, wbc)));
+    }
+
+    @FunctionalInterface
+    private static interface ThrowingLongSupplier<E extends Exception> {
+        long supply() throws E;
+    }
+
+    private static long supplyUsingBlockingChannel(Channel c, ThrowingLongSupplier<IOException> tls) throws IOException {
+        if (c instanceof SelectableChannel sc) {
+            synchronized (sc.blockingLock()) {
+                if (!sc.isBlocking()) {
+                    throw new IllegalBlockingModeException();
+                }
+                return tls.supply();
```

Comment:
It does not look like the `SelectableChannel` branch is tested, including whether an `IllegalBlockingModeException` is thrown when it should be.